lib/gc/Transforms/BufferHoist.cpp

```diff
@@ -0,0 +1,433 @@
+          // runtime var
+          Value result =
+              builder.create<arith::MulIOp>(loc, originShapeValue, inductVar);
+          offset[0] = result;
+
+          SmallVector<OpFoldResult> strides(memrefTypeRank,
+                                            builder.getI64IntegerAttr(1));
+
+          auto subView = builder.create<memref::SubViewOp>(
+              allocOperation->getLoc(), newAlloc, offset, size, strides);
+          allocOperation->replaceAllUsesWith(subView->getResults());
+          allocOperation->remove();
+        } else {
+          return failure();
+        }
+      } else {
+        allocOperation->moveBefore(startOperation);
+      }
+    }
+    return success();
+  }
+
+private:
+  /// Finds a valid placement block by walking upwards in the CFG until we
+  /// either cannot continue our walk due to constraints (given by the StateT
+  /// implementation) or we have reached the upper-most dominator block.
+  Block *findPlacementBlock(StateT &state, Block *upperBound,
+                            int32_t hoistLevel = 0) {
+    Block *currentBlock = state.placementBlock;
+    // ppropriate placement block that satisfies the constraint of the
+    // current StateT implementation. Walk until we reach the upperBound
+    // block (if any).
+
+    // If we are not able to find a valid parent operation or an
+    // associated parent block, break the walk loop.
+    Operation *parentOp;
+    Block *parentBlock;
+    while ((parentOp = currentBlock->getParentOp()) &&
+           (parentBlock = parentOp->getBlock()) &&
+           (!upperBound ||
+            dominators.properlyDominates(upperBound, currentBlock))) {
+      // Try to find an immediate dominator and check whether the parent block
+      // is above the immediate dominator (if any).
+      DominanceInfoNode *idom = nullptr;
+
+      // DominanceInfo doesn't support getNode queries for single-block regions.
+      if (!currentBlock->isEntryBlock())
+        idom = dominators.getNode(currentBlock)->getIDom();
+
+      if (idom && dominators.properlyDominates(parentBlock, idom->getBlock())) {
+        // If the current immediate dominator is below the placement block, move
+        // to the immediate dominator block.
+        currentBlock = idom->getBlock();
+        state.recordMoveToDominator(currentBlock);
+      } else {
+        // We have to move to our parent block since an immediate dominator does
+        // either not exist or is above our parent block. If we cannot move to
+        // our parent operation due to constraints given by the StateT
+        // implementation, break the walk loop. Furthermore, we should not move
+        // allocations out of unknown region-based control-flow operations.
+        if ((!isKnownControlFlowInterface(parentOp) ||
+             !state.isLegalPlacement(parentOp)) &&
+            (hoistLevel == 0))
+          break;
+        if (isa<scf::ForallOp>(parentOp) && hoistLevel > 0)
+          hoistLevel--;
```

Comment:
"Why is the variable 'hoistLevel' being decreased by 1? This variable is passed by value."

Reply:
To make sure this thread-local allocator will be hoisted to the outer-most `parallel for` section. For example:
```
scf.forall ...
  scf.forall ...
    scf.forall ...
      %alloc = memref.alloc() ...
```

Reply:
> To make sure this thread-local allocator will be hoisted to the outer-most `parallel for` section. For example:
> 
> ```
> scf.forall ...
>   scf.forall ...
>     scf.forall ...
>       %alloc = memref.alloc() ...
> ```

OK, I meant that the variable 'hoistLevel' is passed by value here, so decrementing it by one does not affect the logic after exiting this function, such as creating a new allocOp or moving the new allocOp in the correct insert point.Therefore, should it actually be passed by reference here?

Reply:
The value of `hoistLevel` is not related to any `allocOp` but only the `scf.forall` Op, we want the original `hoistLevel` value after exiting this private function. So the value instead of a reference is required here.